far/strmix.cpp

```diff
@@ -911,71 +993,33 @@ bool SearchString(string_view const Source, const string& Str, const string& Upp
 	if (!Regexp && PreserveStyle && PreserveStyleReplaceString(Source, Str, ReplaceStr, CurPos, Case, WholeWords, WordDiv, Reverse, *SearchLength))
 		return true;
 
+	const int SourceSize = static_cast<int>(Source.size());
+
 	if (Reverse)
 	{
+		// MZK 2018-04-01 BUGBUG: regex reverse search: "^$" does not match empty string
 		Position--;
 
-		if (Position >= static_cast<int>(Source.size()))
-			Position = static_cast<int>(Source.size() - 1);
+		if (Position >= SourceSize)
+			Position = SourceSize - 1;
 
-		if (Position<0)
+		if (Position < 0)
 			return false;
 	}
 
-	if ((Position < static_cast<int>(Source.size()) || (!Position && Source.empty())) && !Str.empty())
+	if ((Position < SourceSize || (Position == 0 && Source.empty())) && !Str.empty())
 	{
 		if (Regexp)
 		{
-			intptr_t n = re.GetBracketsCount();
-			bool found = false;
-			int half = 0;
-			if (!Reverse)
-			{
-				if (re.SearchEx(Source.raw_data(), Source.raw_data() + Position, Source.raw_data() + Source.size(), pm, n, hm))
-				{
-					found = true;
-				}
-				else
-				{
-					ReMatchErrorMessage(re);
-				}
-			}
-			else
-			{
-				int pos = 0;
-				for (;;)
-				{
-					if (!re.SearchEx(Source.raw_data(), Source.raw_data() + pos, Source.raw_data() + Source.size(), pm + half, n, hm))
-					{
-						ReMatchErrorMessage(re);
-						break;
-					}
-					pos = static_cast<int>(pm[half].start);
-					if (pos > Position)
-						break;
-
-					found = true;
-					++pos;
-					half = n - half;
-				}
-				half = n - half;
-			}
-			if (found)
-			{
-				*SearchLength = pm[half].end - pm[half].start;
-				CurPos = pm[half].start;
-				ReplaceStr = ReplaceBrackets(Source, ReplaceStr, pm + half, n, hm);
-			}
-
-			return found;
+			return SearchStringRegex(Source, re, pm, hm, Position, Reverse, ReplaceStr, CurPos, SearchLength);
```

Comment:
The refactoring was mostly extracting of the `SearchStringRegex` helper function (plus some shuffling of things around). Unfortunately, the diff is not great.